### PR TITLE
Revamp simplerunner.lua to make running/creating benchmarks easier 

### PR DESCRIPTION
**Pull request by fsfod**
This is revamp of simplerunner.lua to help achieve the goal of #23 for providing a way to quickly run benchmarks and help in the process of making them.
There is now a basic command line option system borrowed from dynasm.lua. Some basic stats are now calculated and printed from all the iteration times. 
Passing no command line augments to simplerunner will cause it to run all the benchmarks each for 30 iterations and a default scaling of one. To run a single benchmark just pass benchmark name as the last command line entry. An optional iteration count can be specified after it like so

`simplerunner.lua luafun 1500`

Added basic progress indicator by printing a dot to the console for each iteration run.
Made collecting and printing jitstats off by default now..
Turned off JITing benchmark loop to try to keep the timing more stable.
Added optional support for using a more high res timer by using the ffi system otherwise os.clock is used. Also the functions used to get the highres time are also pre JIT'ed to try keep the timing stable.

### CODE REVIEW

File: lualibs/time.lua [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+--wall-clock time, monotonic time and sleeping for Windows, Linux and OSX.
+--Written by Cosmin Apreutesei. Public Domain.
+
+local ffi = require'ffi'
+
+local M = {}
+local C = ffi.C
+
+if ffi.os == 'Windows' then
+
```
Comment (ltratt) [inline lualibs/time.lua]: Blank line... and this file uses tabs for indentation?

File: simplerunner.lua [hunk 2]
```diff
@@ -31,30 +56,51 @@ local benchlist = {
...
+if timeloaded then
+    clock = time.clock
+ 
+    if jit_util then
+        -- Force the timer functions to get JIT'ed so they don't go through the slow ffi path in the interpreter thats data driven
+        jitfunc(startimer)
+        jitfunc(stoptimer)
     end
+else
+    -- Uses os.clock uses the libc 'clock' function. CLOCKS_PER_SEC is normally 1Mhz on Linux but on windows its was 1khz.
```
Comment (ltratt) [inline simplerunner.lua]: "Uses" twice?

File: simplerunner.lua [hunk 3]
```diff
@@ -63,53 +109,276 @@ function runbench(name, count)
...
-        local start = os.clock()
-        run_iter(count)
-        print(name.." took", os.clock() - start)
+       times = {}
     end
+    io.write("Running "..name..": ")
+    io.flush()
 
+    local run_iter = _G.run_iter
+    for i= 1, count do
```
Comment (ltratt) [inline simplerunner.lua]: Missing space before `=`?

File: simplerunner.lua [hunk 4]
```diff
@@ -63,53 +109,276 @@ function runbench(name, count)
...
+    
+    if loading_jitstats and jitstats.stats.starts > 0 then 
+        jitstats.print("Traces created while loading benchmark")
+        jitstats.reset()
+    end  
+end  
 
-    count = count or 1
+function runbench(name, count, scaling)
+    scaling = scaling or 1
```
Comment (ltratt) [inline simplerunner.lua]: I'm not sure I see the point of this line.

File: simplerunner.lua [hunk 5]
```diff
@@ -63,53 +109,276 @@ function runbench(name, count)
...
+    loadbench(name)
```
Comment (ltratt) [inline simplerunner.lua]: Can we unload the benchmark at the end of the benchmarking?

File: simplerunner.lua [hunk 6]
```diff
@@ -63,53 +109,276 @@ function runbench(name, count)
...
+
+-- Short aliases for long options.
+local opt_alias = {
+    h = "help", ["?"] = "help", V = "version",
+    b = "bench", c = "count", s = "scaling",
+}
+
+-- Print help text.
+function opt_map.help()
+  io.stdout:write[[
```
Comment (ltratt) [inline simplerunner.lua]: Two space indent?

File: simplerunner.lua [hunk 7]
```diff
@@ -63,53 +109,276 @@ function runbench(name, count)
...
+-- Misc. options.
+function opt_map.count(args) g_opt.count = optparam(args, "number") end
+function opt_map.bench(args) g_opt.bench = optparam(args) end
+function opt_map.scaling(args) g_opt.scaling = optparam(args, "number") end
+function opt_map.jitstats() g_opt.jitstats = true end
+function opt_map.benchloadjitstats() loading_jitstats = true end
+function opt_map.nogc() g_opt.nogc = true end
+function opt_map.jitlog() g_opt.jitlog = true end
+
+------------------------------------------------------------------------------
```
Comment (ltratt) [inline simplerunner.lua]: Not sure this line is helpful.

**Comment by ltratt** [comment]
Does Lua not have a command-line parameter parsing library? Writing our own feels a bit retrograde!

File: simplerunner.lua [hunk 8]
```diff
@@ -342,25 +342,22 @@ local function parseargs(args)
...
     local nargs = #args - args.argn + 1
-    if not g_opt.bench and g_opt.count then
-        if nargs == 0 then
-            -- Show help and exit
-            opt_map.help()
+    if (not g_opt.bench or not g_opt.count) and nargs > 0 then
+        local arg1 = args[args.argn]
+        local num = tonumber(arg1)
+        if num then
+            -- Just a iteration count was passed ignore any extra args after
```
Comment (ltratt) [inline simplerunner.lua]: "aN iteration count". Missing ";" before "ignore"?

File: simplerunner.lua [hunk 9]
```diff
@@ -342,25 +342,22 @@ local function parseargs(args)
...
-            else
-                local arg2 = args[args.argn + 1]
-                g_opt.bench = arg1
-                if arg2 then
-                    num = tonumber(arg2)
-                    if not num then
-                        opterror("Bad count value '", arg2, "' to run benchmarks for")
-                    end
-                    g_opt.count = num
+            -- We expect a benchmark name followed by a optional iteration count
```
Comment (ltratt) [inline simplerunner.lua]: "aN" optional.

File: lualibs/time.lua (hunk 1 above)
Comment (fsfod) [reply lualibs/time.lua]: This a Lua external module that i didn't make

Comment (ltratt) [reply lualibs/time.lua]: Why are we copying it into our repo and not downloading it in a build script? We don't want to maintain forks of such things if we can avoid it...

Comment (fsfod) [reply lualibs/time.lua]: Its not a Lua rocks module https://github.com/luapower/time so there is no a package manager to get it from.

File: simplerunner.lua (hunk 5 above)
Comment (fsfod) [reply simplerunner.lua]: well i do nil out the run_iter global function but that doesn't really do much unless we run the garbage collector

File: lualibs/time.lua (hunk 1 above)
Comment (ltratt) [reply lualibs/time.lua]: We could just download it via this link `https://raw.githubusercontent.com/luapower/time/master/time.lua`.

File: simplerunner.lua (hunk 5 above)
Comment (ltratt) [reply simplerunner.lua]: I think it would be fine to run the garbage collector after an unload. Running multiple benchmarks in the same process has weird effects; anything we can do to lessen those effects is a good idea (even though I know we can't do *that* much).

**Comment by ltratt** [comment]
I have a simple feature request to stick in this PR: let's randomise the order of benchmarks whenever we run them (even if the user specifies benchmarks, we'll randomise that order). That way, if you run the benchmark suite twice, you're less likely to be caught out by the effects that one benchmark might have on the next. [This is an imperfect mitigation strategy, but it's better than nothing.]

**Comment by vext01** [comment]
Good idea.

File: lualibs/time.lua (hunk 1 above)
Comment (ltratt) [reply lualibs/time.lua]: I think everyone can get hold of `curl`/`wget` and we should assume that someone capable of using this repository knows how to do so. Duplicating code is just a recipe for disaster long-term: it means we don't get bug fixes etc.

**Comment by fsfod** [comment]
> Does Lua not have a command-line parameter parsing library? Writing our own feels a bit retrograde!
Lua only has a minimal standard library

> Can we run things on Travis for less time to save them some energy. I'd have thought a parameter of 2 or 3 would be sufficient.
I already reduced the time benchmarks run for travis. Most of the time is spent building all the flavours of Lua. We also don't have any scaling values setup in this runner

File: simplerunner.lua (hunk 4 above)
Comment (fsfod) [reply simplerunner.lua]: Its defaulting it to 1 or if the parameter is unset

Comment (ltratt) [reply simplerunner.lua]: I can't see anywhere where the parameter is unset though?

File: simplerunner.lua (hunk 7 above)
Comment (fsfod) [reply simplerunner.lua]: its was there from the code i copied it from although the ordering is slightly different

Comment (ltratt) [reply simplerunner.lua]: Copied from where? What's the license of the thing that was copied?

File: simplerunner.lua (hunk 4 above)
Comment (fsfod) [reply simplerunner.lua]: well the run_all_benchmarks call at the bottom doesn't pass it and i think passing a scaling value to run_all_benchmarks would be an override from a preconfigured scaling settings system we don't have

Comment (ltratt) [reply simplerunner.lua]: I'm not sure I understand the sentence after "i think passing" -- can you rephrase?

Comment (fsfod) [reply simplerunner.lua]: I think we should have our benchmark scaling values in a shared json file used by both our krun file and this

Comment (ltratt) [reply simplerunner.lua]: I'm fine with that, at which point we wouldn't need this default?

**Comment by fsfod** [comment]
@ltratt I've added benchmark randomization, garbage collection after each benchmark and per benchmark scaling from a json config file.

File: simplerunner.lua [hunk 10]
```diff
@@ -95,7 +95,7 @@ if timeloaded then
         jitfunc(stoptimer)
     end
 else
-    -- Uses os.clock uses the libc 'clock' function. CLOCKS_PER_SEC is normally 1Mhz on Linux but on windows its was 1khz.
+    -- Fallback to os.clock this uses the libc 'clock' function. CLOCKS_PER_SEC is normally 1Mhz on Linux but on windows its was 1khz.
```
Comment (ltratt) [inline simplerunner.lua]: "Fallback to os.clock WHICH uses".

File: simplerunner.lua [hunk 11]
```diff
@@ -167,10 +167,25 @@ if jit then
     jit.off(runbench)
 end
 
-function run_all_benchmarks(count, scaling)
-    assert(#benchlist ~= 0)
+function permute(tab)
+    local count = #tab
+    math.randomseed(os.time())
+    for i, v in ipairs(tab) do
+        local j = math.random(i, count)
```
[the remark this answers is not in the document]
Comment (ltratt) [reply simplerunner.lua]: I don't think it's quite equivalent: look at the upper bound, for example. [I think you've almost, but not quite, implemented the version Wikipedia describes as "An equivalent version which shuffles the array in the opposite direction (from lowest index to highest) is:"]

**Comment by fsfod** [comment]
@ltratt i fixed the shuffle and various little things

**Comment by ltratt** [comment]
Thanks. If you squash, I can merge.

**Comment by fsfod** [comment]
@ltratt squashed and ready to merge
